File: Opsec/Opsec.java

// test794d : USAG-Lib opsec
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.zip.CRC32;

/*
Opsec header handler, !!! DO NOT REUSE THIS OBJECT !!! reset after reading body key
pw: (msg), headAlgo, salt, pwHash, encHeadData
rsa: (msg), headAlgo, encHeadKey, encHeadData
ecc: (msg), headAlgo, encHeadData
header: (smsg), (size), (name), (bodyKey), (bodyAlgo), (contAlgo), (sign)
*/
public class Opsec {
    // Outer Layer
    public String Msg; // non-secured message
    private String headAlgo; // header algorithm, [arg1 pbk1 rsa1 ecc1]
    private byte[] salt; // salt
    private byte[] pwHash; // pw hash
    private byte[] encHeadKey; // encrypted header key
    private byte[] encHeadData; // encrypted header data

    // Inner Layer
    public String Smsg; // secured message
    public long Size; // full body size, flag for bodyKey generation
    public String Name; // body name
    public byte[] BodyKey; // body key
    public String BodyAlgo; // body algorithm, [gcm1 gcmx1]
    public String ContAlgo; // container algorithm, [zip1 tar1]
    private byte[] sign; // signature to bodyKey/smsg

    public Opsec() {
        Reset();
    }

    public void Reset() {
        Msg = "";
        headAlgo = "";
        salt = new byte[0];
        pwHash = new byte[0];
        encHeadKey = new byte[0];
        encHeadData = new byte[0];

        Smsg = "";
        Size = -1;
        Name = "";
        BodyKey = new byte[0];
        BodyAlgo = "";
        ContAlgo = "";
        sign = new byte[0];
    }

    // ========== Helper Functions ==========
    public static String Crc32(byte[] data) {
        CRC32 crc = new CRC32();
        crc.update(data);
        long value = crc.getValue();
        return String.format("%02x%02x%02x%02x", (value & 0xFF), (value >> 8) & 0xFF, (value >> 16) & 0xFF,
                (value >> 24) & 0xFF); // 8 chars hex string
    }

    public byte[] EncodeInt(long data, int size) {
        ByteBuffer buf = ByteBuffer.allocate(size).order(ByteOrder.LITTLE_ENDIAN);
        if (size == 1)
            buf.put((byte) data);
        else if (size == 2)
            buf.putShort((short) data);
        else if (size == 4)
            buf.putInt((int) data);
        else if (size == 8)
            buf.putLong(data);
        return buf.array();
    }

    public long DecodeInt(byte[] data) {
        ByteBuffer buf = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN);
        if (data.length == 1)
            return Byte.toUnsignedInt(buf.get());
        if (data.length == 2)
            return Short.toUnsignedInt(buf.getShort());
        if (data.length == 4)
            return Integer.toUnsignedLong(buf.getInt());
        if (data.length == 8)
            return buf.getLong(); // Java long is signed, but bits are same
        return 0;
    }

    private byte[] concat(byte[]... arrays) {
        int len = 0;
        for (byte[] a : arrays)
            len += a.length;
        byte[] res = new byte[len];
        int pos = 0;
        for (byte[] a : arrays) {
            System.arraycopy(a, 0, res, pos, a.length);
            pos += a.length;
        }
        return res;
    }

    private byte[] strToBytes(String s) {
        return s.getBytes(StandardCharsets.UTF_8);
    }

    private String bytesToStr(byte[] b) {
        return new String(b, StandardCharsets.UTF_8);
    }

    // Config Encoding
    public byte[] EncodeCfg(Map<String, byte[]> data) throws IOException {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        for (Map.Entry<String, byte[]> entry : data.entrySet()) {
            byte[] keyBytes = strToBytes(entry.getKey());
            byte[] valBytes = entry.getValue();
            int keyLen = keyBytes.length;
            int dataLen = valBytes.length;
            if (keyLen > 127)
                throw new IllegalArgumentException("Key length too long: " + keyLen);
            if (dataLen > 65535)
                throw new IllegalArgumentException("Data size too big: " + dataLen);

            if (dataLen > 255) {
                out.write(keyLen + 128);
                out.write(keyBytes);
                out.write(EncodeInt(dataLen, 2));
            } else {
                out.write(keyLen);
                out.write(keyBytes);
                out.write(dataLen);
            }
            out.write(valBytes);
        }
        return out.toByteArray();
    }

    // Config Decoding
    public Map<String, byte[]> DecodeCfg(byte[] data) {
        Map<String, byte[]> result = new HashMap<>();
        ByteBuffer buf = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN);
        while (buf.hasRemaining()) {
            int keyLen = Byte.toUnsignedInt(buf.get());
            boolean isLongData = false;
            if (keyLen > 127) {
                keyLen -= 128;
                isLongData = true;
            }
            byte[] keyBytes = new byte[keyLen];
            buf.get(keyBytes);
            String key = bytesToStr(keyBytes);

            int dataLen;
            if (isLongData) {
                dataLen = Short.toUnsignedInt(buf.getShort());
            } else {
                dataLen = Byte.toUnsignedInt(buf.get());
            }
            byte[] valBytes = new byte[dataLen];
            buf.get(valBytes);
            result.put(key, valBytes);
        }
        return result;
    }

    // read stream, return opsec header
    public byte[] Read(InputStream ins, int cut) throws IOException {
        int c = 0;
        while (true) {
            byte[] buf4 = ins.readNBytes(4);
            c += 4;

            if (Arrays.equals(buf4, strToBytes("YAS2"))) {
                byte[] buf2 = ins.readNBytes(2);
                long size = DecodeInt(buf2);
                if (size == 65535) {
                    buf2 = ins.readNBytes(2);
                    size += DecodeInt(buf2);
                }
                byte[] packet = new byte[(int) size];
                int totalRead = 0;
                while (totalRead < size) {
                    int r = ins.readNBytes(packet, totalRead, (int) size - totalRead);
                    if (r == -1 || (r == 0 && totalRead < size))
                        throw new java.io.EOFException("Unexpected EOF while reading opsec header");
                    totalRead += r;
                }
                return packet;

            } else {
                ins.readNBytes(124);
                c += 124;
            }
            if (cut > 0 && c > cut)
                return new byte[0];
        }
    }

    // write opsec header to stream
    public void Write(OutputStream outs, byte[] head) throws IOException {
        outs.write(strToBytes("YAS2"));
        int size = head.length;
        if (size < 65535) {
            outs.write(EncodeInt(size, 2));
        } else if (size <= 65535 * 2) {
            outs.write(EncodeInt(65535, 2));
            outs.write(EncodeInt(size - 65535, 2));
        } else {
            throw new IOException("Data size too big: " + size);
        }
        outs.write(head);
    }

    private byte[] wrapHead() throws IOException {
        Map<String, byte[]> cfg = new HashMap<>();
        if (!Smsg.isEmpty())
            cfg.put("smsg", strToBytes(Smsg));
        if (Size >= 0) {
            if (Size < 65536)
                cfg.put("sz", EncodeInt(Size, 2));
            else if (Size < 4294967296L)
                cfg.put("sz", EncodeInt(Size, 4));
            else
                cfg.put("sz", EncodeInt(Size, 8));
        }
        if (!Name.isEmpty())
            cfg.put("nm", strToBytes(Name));
        if (BodyKey.length > 0)
            cfg.put("bkey", BodyKey);
        if (!BodyAlgo.isEmpty())
            cfg.put("bodyal", strToBytes(BodyAlgo));
        if (!ContAlgo.isEmpty())
            cfg.put("contal", strToBytes(ContAlgo));
        if (sign.length > 0)
            cfg.put("sgn", sign);
        return EncodeCfg(cfg);
    }

    private void unwrapHead(byte[] data) {
        Map<String, byte[]> cfg = DecodeCfg(data);
        if (cfg.containsKey("smsg"))
            Smsg = bytesToStr(cfg.get("smsg"));
        if (cfg.containsKey("sz"))
            Size = DecodeInt(cfg.get("sz"));
        if (cfg.containsKey("nm"))
            Name = bytesToStr(cfg.get("nm"));
        if (cfg.containsKey("bkey"))
            BodyKey = cfg.get("bkey");
        if (cfg.containsKey("bodyal"))
            BodyAlgo = bytesToStr(cfg.get("bodyal"));
        if (cfg.containsKey("contal"))
            ContAlgo = bytesToStr(cfg.get("contal"));
        if (cfg.containsKey("sgn"))
            sign = cfg.get("sgn");
    }

    // encrypt with password
    public byte[] Encpw(String method, byte[] pw, byte[] kf) throws Exception {
        Bencrypt worker = new Bencrypt();
        headAlgo = method;
        salt = worker.Random(16);
        if (Size >= 0) {
            BodyKey = worker.Random(44);
        }
        byte[] combinedPw = (kf == null || kf.length == 0) ? pw : concat(pw, kf);
        byte[] mkey;
        String verifyLbl, keygenLbl;

        // generate password hash, encrypt header
        if (method.equals("sha3")) {
            mkey = Bencrypt.SHA3512(concat(salt, combinedPw));
            verifyLbl = "PWHASH_OPSEC_SHA3512";
            keygenLbl = "KEYGEN_OPSEC_SHA3512";
        } else if (method.equals("pbk1")) {
            mkey = Bencrypt.pbkdf2(combinedPw, salt, 1000000, 64);
            verifyLbl = "PWHASH_OPSEC_PBKDF2";
            keygenLbl = "KEYGEN_OPSEC_PBKDF2";
        } else if (method.equals("arg1")) {
            String hashStr = worker.argon2Hash(combinedPw, salt);
            mkey = strToBytes(hashStr);
            verifyLbl = "PWHASH_OPSEC_ARGON2";
            keygenLbl = "KEYGEN_OPSEC_ARGON2";
        } else {
            throw new IllegalArgumentException("Unsupported method: " + method);
        }
        pwHash = Bencrypt.genkey(mkey, verifyLbl, 32);
        byte[] hkey = Bencrypt.genkey(mkey, keygenLbl, 44);

        // Encrypt Header using SymMaster
        byte[] headData = wrapHead();
        Bencrypt.SymMaster sm = new Bencrypt.SymMaster("gcm1", hkey);
        encHeadData = sm.EnBin(headData);

        // wrap header
        Map<String, byte[]> cfg = new HashMap<>();
        if (!Msg.isEmpty())
            cfg.put("msg", strToBytes(Msg));
        cfg.put("headal", strToBytes(headAlgo));
        cfg.put("salt", salt);
        cfg.put("pwh", pwHash);
        cfg.put("ehd", encHeadData);
        return EncodeCfg(cfg);
    }

    // encrypt with public key, sign if private key is not null
    public byte[] Encpub(String method, byte[] publicBytes, byte[] privateBytes) throws Exception {
        Bencrypt worker = new Bencrypt();
        headAlgo = method;
        if (Size >= 0) {
            BodyKey = worker.Random(44);
        }

        // Init Master & Sign
        Bencrypt.AsymMaster am = new Bencrypt.AsymMaster(method);
        am.Loadkey(publicBytes, privateBytes);

        if (privateBytes != null) {
            byte[] s = (BodyKey.length > 0) ? BodyKey : strToBytes(Smsg);
            sign = am.Sign(s);
        }

        // encrypt header
        byte[] headData = wrapHead();
        if (method.equals("rsa1") || method.equals("rsa2")) {
            // RSA Hybrid: Encrypt Key with RSA, Data with AES
            byte[] hkey = worker.Random(44);
            encHeadKey = am.Encrypt(hkey);

            Bencrypt.SymMaster sm = new Bencrypt.SymMaster("gcm1", hkey);
            encHeadData = sm.EnBin(headData);
        } else if (method.equals("ecc1")) {
            // ECC Hybrid: Handled internally by ECC1 class
            encHeadData = am.Encrypt(headData);
        } else {
            throw new IllegalArgumentException("Unsupported method: " + method);
        }

        // wrap header
        Map<String, byte[]> cfg = new HashMap<>();
        if (!Msg.isEmpty())
            cfg.put("msg", strToBytes(Msg));
        cfg.put("headal", strToBytes(headAlgo));
        if (encHeadKey.length > 0)
            cfg.put("ehk", encHeadKey);
        cfg.put("ehd", encHeadData);
        return EncodeCfg(cfg);
    }

    // load outer layer of header
    public void View(byte[] data) {
        Reset();
        Map<String, byte[]> cfg = DecodeCfg(data);
        if (cfg.containsKey("msg"))
            Msg = bytesToStr(cfg.get("msg"));
        if (cfg.containsKey("headal"))
            headAlgo = bytesToStr(cfg.get("headal"));
        if (cfg.containsKey("salt"))
            salt = cfg.get("salt");
        if (cfg.containsKey("pwh"))
            pwHash = cfg.get("pwh");
        if (cfg.containsKey("ehk"))
            encHeadKey = cfg.get("ehk");
        if (cfg.containsKey("ehd"))
            encHeadData = cfg.get("ehd");
    }

    // decrypt with password
    public void Decpw(byte[] pw, byte[] kf) throws Exception {
        Bencrypt worker = new Bencrypt();
        if (headAlgo.isEmpty())
            throw new IllegalStateException("Call view() first");
        byte[] combinedPw = (kf == null || kf.length == 0) ? pw : concat(pw, kf);
        byte[] mkey;
        String verifyLbl, keygenLbl;

        // check password
        if (headAlgo.equals("sha3")) {
            mkey = Bencrypt.SHA3512(concat(salt, combinedPw));
            verifyLbl = "PWHASH_OPSEC_SHA3512";
            keygenLbl = "KEYGEN_OPSEC_SHA3512";
        } else if (headAlgo.equals("pbk1")) {
            mkey = Bencrypt.pbkdf2(combinedPw, salt, 1000000, 64);
            verifyLbl = "PWHASH_OPSEC_PBKDF2";
            keygenLbl = "KEYGEN_OPSEC_PBKDF2";
        } else if (headAlgo.equals("arg1")) {
            String hashStr = worker.argon2Hash(combinedPw, salt);
            mkey = strToBytes(hashStr);
            verifyLbl = "PWHASH_OPSEC_ARGON2";
            keygenLbl = "KEYGEN_OPSEC_ARGON2";
        } else {
            throw new IllegalArgumentException("Unsupported method: " + headAlgo);
        }
        byte[] calcHash = Bencrypt.genkey(mkey, verifyLbl, 32);
        if (!Arrays.equals(calcHash, pwHash))
            throw new SecurityException("Incorrect password");

        // decrypt header
        byte[] hkey = Bencrypt.genkey(mkey, keygenLbl, 44);
        Bencrypt.SymMaster sm = new Bencrypt.SymMaster("gcm1", hkey);
        byte[] decryptedHead = sm.DeBin(encHeadData);
        if (decryptedHead == null)
            throw new SecurityException("AES decryption failed");
        unwrapHead(decryptedHead);
    }

    // decrypt with private key, verify if public key is not null
    public void Decpub(byte[] privateBytes, byte[] publicBytes) throws Exception {
        if (headAlgo.isEmpty())
            throw new IllegalStateException("Call view() first");
        Bencrypt.AsymMaster am = new Bencrypt.AsymMaster(headAlgo);
        am.Loadkey(publicBytes, privateBytes);

        // decrypt header
        byte[] decryptedHead;
        if (headAlgo.equals("rsa1") || headAlgo.equals("rsa2")) {
            byte[] hkey = am.Decrypt(encHeadKey);
            Bencrypt.SymMaster sm = new Bencrypt.SymMaster("gcm1", hkey);
            decryptedHead = sm.DeBin(encHeadData);
        } else if (headAlgo.equals("ecc1")) {
            decryptedHead = am.Decrypt(encHeadData);
        } else {
            throw new IllegalArgumentException("Unsupported method: " + headAlgo);
        }
        if (decryptedHead == null)
            throw new SecurityException("Decryption failed");
        unwrapHead(decryptedHead);

        // verify if public key is not null
        if (publicBytes != null) {
            byte[] s = (BodyKey.length > 0) ? BodyKey : strToBytes(Smsg);
            if (!am.Verify(s, sign)) {
                throw new SecurityException("Signature verification failed");
            }
        }
    }
}

File: Opsec/Opsec.js
// test794b : USAG-Lib opsec
const BencryptURL = 'https://cdn.jsdelivr.net/gh/k-atusa/USAG-Lib/Bencrypt/Bencrypt.js';
const { Random, HashMaster, SymMaster, AsymMaster } = await import(BencryptURL);

// Helper: Fill zeros to buffer
function zeroize(arr) {
    if (arr && arr.byteLength > 0 && typeof arr.fill === 'function') {
        arr.fill(0);
        globalThis.DUMMY_CRYPTO_VOLATILE = arr;
    }
}

// Helper: Concatenate Uint8Arrays
function concat(arrays) {
    let totalLength = 0;
    for (const arr of arrays) {
        totalLength += arr.length;
    }
    const result = new Uint8Array(totalLength);
    let offset = 0;
    for (const arr of arrays) {
        result.set(arr, offset);
        offset += arr.length;
    }
    return result;
}

// Helper: String to Uint8Array / Uint8Array to String
const enc = new TextEncoder();
const dec = new TextDecoder();
function strToU8(str) { return enc.encode(str); }
function u8ToStr(u8) { return dec.decode(u8); }

// Helper: CRC32 Implementation
const crcTable = (() => {
    let c;
    const table = [];
    for (let n = 0; n < 256; n++) {
        c = n;
        for (let k = 0; k < 8; k++) {
            c = ((c & 1) ? (0xEDB88320 ^ (c >>> 1)) : (c >>> 1));
        }
        table[n] = c >>> 0;
    }
    return table;
})();


/**
 * CRC32 Implementation
 * @param {Uint8Array | string} data
 */
export function Crc32(data) {
    const u8 = typeof data === 'string' ? strToU8(data) : data;
    let crc = -1;
    for (let i = 0; i < u8.length; i++) {
        crc = (crc >>> 8) ^ crcTable[(crc ^ u8[i]) & 0xFF];
    }
    crc = (crc ^ (-1)) >>> 0;

    // Return as Hex String (8 chars)
    const buf = new ArrayBuffer(4);
    const view = new DataView(buf);
    view.setUint32(0, crc, true); // little endian
    let hexStr = "";
    for (let i = 0; i < 4; i++) {
        hexStr += view.getUint8(i).toString(16).padStart(2, '0');
    }
    return hexStr;
}

/**
 * Returns size to be padded
 * @param {number} size
 */
export function PadLen(size) {
    if (size <= 0) return 0;

    // 1. 0-16k: 4k*N
    if (size <= 16384) {
        const remainder = size % 4096;
        if (remainder === 0) return 0;
        return 4096 - remainder;
    }

    const bSize = BigInt(size);
    const bitLen = BigInt(bSize.toString(2).length); // get sup bit position
    let k;
    if (bitLen <= 24n) {        // 16k-16m: K=2
        k = 2n;
    } else if (bitLen <= 29n) { // 16m-512m: K=3
        k = 3n;
    } else if (bitLen <= 33n) { // 512m-8g: K=4
        k = 4n;
    } else {                    // 8g+: K=5
        k = 5n;
    }

    // mask and ceiling
    const shift = bitLen - k;
    const mask = (1n << shift) - 1n;
    if ((bSize & mask) === 0n) { // on border size is not padded
        return 0;
    }

    // return actual padding length
    const aftersize = ((bSize >> shift) + 1n) << shift;
    return Number(aftersize - bSize);
}

/**
 * Add padding at the end
 * @param {Object} f 
 * @param {number} size 
 */
export async function PadFile(f, size) {
    for (let i = 0; i < Math.floor(size / 32768); i++) await f.write(Random(32768));
    if (size % 32768 > 0) await f.write(Random(size % 32768));
}

/**
 * Little Endian Integer Encoding
 * @param {number} data 
 * @param {number} size 
 * @returns {Uint8Array}
 */
export function EncodeInt(data, size) {
    const buf = new ArrayBuffer(size);
    const view = new DataView(buf);
    if (size === 1) view.setUint8(0, data);
    else if (size === 2) view.setUint16(0, data, true);
    else if (size === 4) view.setUint32(0, data, true);
    else if (size === 8) view.setBigUint64(0, BigInt(data), true);
    return new Uint8Array(buf);
}

/**
 * Little Endian Integer Decoding
 * @param {Uint8Array} data 
 * @returns {number}
 */
export function DecodeInt(data) {
    const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
    if (data.length === 1) return view.getUint8(0);
    if (data.length === 2) return view.getUint16(0, true);
    if (data.length === 4) return view.getUint32(0, true);
    if (data.length === 8) return Number(view.getBigUint64(0, true)); // Check logic if > safe integer
    return 0;
}

/**
 * Config Encoder, keysize max 127, datasize max 65535
 * @param {Object} data 
 * @returns {Uint8Array}
 */
export function EncodeCfg(data) {
    const chunks = [];
    for (const [key, val] of Object.entries(data)) {
        const valU8 = (typeof val === 'string') ? strToU8(val) : val;
        const keyBytes = strToU8(key);
        const keyLen = keyBytes.length;
        const dataLen = valU8.length;
        if (keyLen > 127) throw new Error(`Key length too long: ${keyLen}`);
        if (dataLen > 65535) throw new Error(`Data size too big: ${dataLen}`);

        if (dataLen > 255) { // DataLen > 255, datasize is 2B, keyLen Flag set (keyLen + 128)
            chunks.push(new Uint8Array([keyLen + 128]));
            chunks.push(keyBytes);
            chunks.push(EncodeInt(dataLen, 2));
        } else { // DataLen <= 255, datasize is 1B
            chunks.push(new Uint8Array([keyLen]));
            chunks.push(keyBytes);
            chunks.push(new Uint8Array([dataLen]));
        }
        chunks.push(valU8);
    }
    return concat(chunks);
}

/**
 * Config Decoder
 * @param {Uint8Array} data 
 * @returns {Object}
 */
export function DecodeCfg(data) {
    const result = {};
    let offset = 0;
    const totalLen = data.length;
    while (offset < totalLen) {
        // get key
        let keyLen = data[offset];
        let isLongData = false;
        offset += 1;
        if (keyLen > 127) {
            keyLen -= 128;
            isLongData = true;
        }
        const keyBytes = data.slice(offset, offset + keyLen);
        const key = u8ToStr(keyBytes);
        offset += keyLen;

        // get data
        let dataLen = 0;
        if (isLongData) {
            const lenBytes = data.slice(offset, offset + 2);
            dataLen = DecodeInt(lenBytes);
            offset += 2;
        } else {
            dataLen = data[offset];
            offset += 1;
        }
        result[key] = data.slice(offset, offset + dataLen);
        offset += dataLen;
    }
    return result;
}

// Opsec header handler
export class Opsec {
    constructor() {
        this.Reset();
        this.SaltLen = 32;
    }

    // reset after reading BodyKey
    Reset() {
        this._headAlgo = "";
        this.Msg = "";
        this.MsgInfo = new Uint8Array(0);

        this._salt = new Uint8Array(0);
        this._pwHash = new Uint8Array(0);
        this._encHeadData = new Uint8Array(0);

        this.Smsg = "";
        this.SmsgInfo = new Uint8Array(0);
        this._sign = new Uint8Array(0);

        this.BodyAlgo = "";
        this.BodyKey = new Uint8Array(0);
        this.BodySize = -1;
        this.BodyInfo = new Uint8Array(0);
    }

    /**
     * read stream, returns Opsec header
     * @param {Object} ins // Object with async read(size)
     * @param {number} cut 
     * @returns {Uint8Array}
     */
    async Read(ins, cut = 65535) {
        let c = 0;
        while (true) {
            const data = await ins.read(4);
            c += 4;
            if (data.length === 0) return new Uint8Array(0);

            const magic = u8ToStr(data);
            if (magic === "YAS2") {
                const sizeBuf = await ins.read(2);
                let size = DecodeInt(sizeBuf);
                if (size === 65535) {
                    const extSizeBuf = await ins.read(2);
                    size += DecodeInt(extSizeBuf);
                }
                return await ins.read(size);
            } else {
                await ins.read(124);
                c += 124;
            }
            if (cut > 0 && c > cut) return new Uint8Array(0);
        }
    }

    /**
     * write opsec header to stream
     * @param {Object} outs // Object with async write(data)
     * @param {Uint8Array} head
     */
    async Write(outs, head) {
        await outs.write(strToU8("YAS2"));
        const size = head.length;
        if (size < 65535) {
            await outs.write(EncodeInt(size, 2));
        } else if (size <= 65535 * 2) {
            await outs.write(EncodeInt(65535, 2));
            await outs.write(EncodeInt(size - 65535, 2));
        } else {
            throw new Error(`Data size too big: ${size}`);
        }
        await outs.write(head);
    }

    _wrapEncHead() {
        const cfg = {};
        if (this.Smsg !== "") cfg["smsg"] = this.Smsg;
        if (this.SmsgInfo.length > 0) cfg["sinf"] = this.SmsgInfo;
        if (this._sign.length > 0) cfg["sgn"] = this._sign;
        if (this.BodyAlgo !== "") cfg["bal"] = this.BodyAlgo;
        if (this.BodyKey.length > 0) cfg["bkey"] = this.BodyKey;
        if (this.BodySize >= 0) {
            if (this.BodySize < 65536) cfg["bsz"] = EncodeInt(this.BodySize, 2);
            else if (this.BodySize < 4294967296) cfg["bsz"] = EncodeInt(this.BodySize, 4);
            else cfg["bsz"] = EncodeInt(this.BodySize, 8);
        }
        if (this.BodyInfo.length > 0) cfg["binf"] = this.BodyInfo;
        return EncodeCfg(cfg);
    }

    _unwrapEncHead(data) {
        const cfg = DecodeCfg(data);
        if (cfg["smsg"]) this.Smsg = u8ToStr(cfg["smsg"]);
        if (cfg["sinf"]) this.SmsgInfo = cfg["sinf"];
        if (cfg["sgn"]) this._sign = cfg["sgn"];
        if (cfg["bal"]) this.BodyAlgo = u8ToStr(cfg["bal"]);
        if (cfg["bkey"]) this.BodyKey = cfg["bkey"];
        if (cfg["bsz"]) this.BodySize = DecodeInt(cfg["bsz"]);
        if (cfg["binf"]) this.BodyInfo = cfg["binf"];
    }

    /**
     * Encrypt with password, returns header
     * @param {string} method 
     * @param {Uint8Array} pw 
     * @param {Uint8Array} kf 
     * @returns {Uint8Array}
     */
    async Encpw(method, pw, kf = new Uint8Array(0)) {
        // generate random parameters
        this._headAlgo = method;
        this._salt = Random(this.SaltLen);
        if (this.BodySize >= 0) {
            this.BodyKey = Random(44);
        }

        const pwBytes = (typeof pw === 'string') ? strToU8(pw) : pw;
        const kfBytes = (typeof kf === 'string') ? strToU8(kf) : kf;
        const combinedPw = concat([pwBytes, kfBytes]);

        // get pwhash & header key, encrypt header
        const hm = new HashMaster(method);
        const [pwHash, hkey] = await hm.KDF(combinedPw, this._salt);
        zeroize(combinedPw);
        this._pwHash = pwHash;

        const headData = this._wrapEncHead();
        const sm = new SymMaster("gcm1", hkey);
        this._encHeadData = await sm.EnBin(headData);
        zeroize(headData);
        zeroize(hkey);

        // wrap header
        const cfg = {};
        if (this.Msg !== "") cfg["msg"] = this.Msg;
        if (this.MsgInfo.length > 0) cfg["minf"] = this.MsgInfo;
        cfg["hal"] = this._headAlgo;
        cfg["salt"] = this._salt;
        cfg["pwh"] = this._pwHash;
        cfg["ehd"] = this._encHeadData;
        return EncodeCfg(cfg);
    }

    /**
     * Encrypt with public key, returns header
     * @param {string} method 
     * @param {Uint8Array} peerPub 
     * @param {Uint8Array|null} myPri 
     * @returns {Uint8Array}
     */
    async Encpub(method, peerPub, myPri = null) {
        // generate random parameters
        this._headAlgo = method;
        if (this.BodySize >= 0) {
            this.BodyKey = Random(44);
        }

        const peerPubBytes = (typeof peerPub === 'string') ? strToU8(peerPub) : peerPub;

        // sign with private key if provided
        if (myPri !== null) {
            const am = new AsymMaster(method);
            await am.Loadkey(null, myPri);
            // sign to [hal][peerPub][smsg][sinf]
            const signTgt = concat([
                strToU8(method),
                peerPubBytes,
                strToU8(this.Smsg),
                this.SmsgInfo
            ]);
            this._sign = await am.Sign(signTgt);
            zeroize(signTgt);
        }

        // encrypt header
        const am = new AsymMaster(method);
        await am.Loadkey(peerPubBytes, null);
        const headData = this._wrapEncHead();

        if (method === "rsa1" || method === "rsa2") {
            // RSA Hybrid: Encrypt Key with RSA, Data with AES
            const hkey = Random(44);
            this.MsgInfo = await am.Encrypt(hkey);
            const sm = new SymMaster("gcm1", hkey);
            this._encHeadData = await sm.EnBin(headData);
            zeroize(hkey);
        } else {
            this._encHeadData = await am.Encrypt(headData);
        }
        zeroize(headData);

        // wrap header
        const cfg = {};
        if (this.Msg !== "") cfg["msg"] = this.Msg;
        if (this.MsgInfo.length > 0) cfg["minf"] = this.MsgInfo;
        cfg["hal"] = this._headAlgo;
        cfg["ehd"] = this._encHeadData;
        return EncodeCfg(cfg);
    }

    /**
     * load outer layer of header
     * @param {Uint8Array} data
     */
    View(data) {
        this.Reset();
        const cfg = DecodeCfg(data);
        if (cfg["msg"]) this.Msg = u8ToStr(cfg["msg"]);
        if (cfg["minf"]) this.MsgInfo = cfg["minf"];
        if (cfg["hal"]) this._headAlgo = u8ToStr(cfg["hal"]);
        if (cfg["salt"]) this._salt = cfg["salt"];
        if (cfg["pwh"]) this._pwHash = cfg["pwh"];
        if (cfg["ehd"]) this._encHeadData = cfg["ehd"];
    }

    /**
     * Decrypt with password
     * @param {Uint8Array} pw 
     * @param {Uint8Array} kf 
     */
    async Decpw(pw, kf = new Uint8Array(0)) {
        if (this._headAlgo === "") throw new Error("Call view() first");
        const pwBytes = (typeof pw === 'string') ? strToU8(pw) : pw;
        const kfBytes = (typeof kf === 'string') ? strToU8(kf) : kf;
        const combinedPw = concat([pwBytes, kfBytes]);

        // check parameters, get header key
        const hm = new HashMaster(this._headAlgo);
        const [pwHash, hkey] = await hm.KDF(combinedPw, this._salt);
        zeroize(combinedPw);

        // check password (constant time comparison)
        if (pwHash.length !== this._pwHash.length) throw new Error("Incorrect password");
        let diff = 0;
        for (let i = 0; i < pwHash.length; i++) {
            diff |= pwHash[i] ^ this._pwHash[i];
        }
        if (diff !== 0) throw new Error("Incorrect password");

        // decrypt header
        const sm = new SymMaster("gcm1", hkey);
        const headData = await sm.DeBin(this._encHeadData);
        zeroize(hkey);
        this._unwrapEncHead(headData);
        zeroize(headData);
    }

    /**
     * Decrypt with private key
     * @param {Uint8Array} myPri 
     * @param {Uint8Array|null} myPub 
     * @param {Uint8Array|null} peerPub
     */
    async Decpub(myPri, myPub = null, peerPub = null) {
        // check parameters, decrypt header
        if (this._headAlgo === "") throw new Error("Call view() first");
        const am = new AsymMaster(this._headAlgo);
        await am.Loadkey(null, myPri);

        let headData;
        if (this._headAlgo === "rsa1" || this._headAlgo === "rsa2") {
            // RSA Hybrid
            const hkey = await am.Decrypt(this.MsgInfo);
            const sm = new SymMaster("gcm1", hkey);
            headData = await sm.DeBin(this._encHeadData);
            zeroize(hkey);
        } else {
            headData = await am.Decrypt(this._encHeadData);
        }
        this._unwrapEncHead(headData);
        zeroize(headData);

        // verify sign
        if (myPub === null && peerPub === null) return;
        if (myPub === null || peerPub === null) {
            if (this._sign.length > 0) throw new Error("Both myPub and peerPub should be provided to verify sign");
            return;
        }

        const amVerify = new AsymMaster(this._headAlgo);
        await amVerify.Loadkey(peerPub, null);
        const signTgt = concat([
            strToU8(this._headAlgo),
            myPub,
            strToU8(this.Smsg),
            this.SmsgInfo
        ]);

        const verified = await amVerify.Verify(signTgt, this._sign);
        if (!verified) throw new Error("Sign verification failed");
    }
}